refactor: New attempt at reset

lukelooker opened the pull request:
Work in progress, please ignore


## Developer Checklist [ℹ️](https://github.com/looker-open-source/components/blob/main//CONTRIBUTING.md#developer-checklist)

- [ ] ♿️ Accessibility addressed
- [ ] 🌏 Localization & Internationalization addressed
- [ ] 🖼 Image Snapshot coverage
- [ ] 📚 Documentation updated
- [ ] 🧳 Bundle size impact addressed
- [ ] 🏁 Performance impacts addressed
- [ ] 👾 Browsers tested
  - [ ] Chrome
  - [ ] Firefox
  - [ ] Safari
  - [ ] IE11

lukelooker commented on packages/components-providers/src/ComponentsProvider.tsx:
```diff
@@ -59,10 +52,33 @@ export interface ComponentsProviderProps
    */
+  snapshotMode?: boolean
+
+  /**
+   * Prevent automatic injection of a basic CSS-reset into the DOM
+   * @deprecated - no longer has any actual effect. Global reset no longer in use.
+   * @todo - Remove for 2.x series
+   */
+  globalStyle?: boolean
+
+  /**
+   * Enable style support for IE11
+   * @deprecated - no longer has any actual effect, IE11 support no longer requires
+   *  specialized implementation with changes to reset implementation
+   * @todo - Remove for 2.x series
+   */
   ie11Support?: boolean
```
Keeping these to preserve backwards compatibility (even though they no longer _actually_ do anything)

lukelooker commented on packages/components-providers/src/StyleDefender.tsx:
```diff
@@ -24,48 +24,33 @@
-
-export const GlobalStyle = createGlobalStyle`
-  ${reset}
+  /**
+    * Global reset applied to prevent styling on top level tags outside of Looker
+    * UI Components from interfering with styles.
+    **/
+  ${({ theme: { reset } }) => reset && reset()}
```
Debating whether or not it makes sense to keep this in place. It probably opens up some "unique" opportunities for trouble - arguably we could remove it from here (but keep it in the theme) and remove the feature from theme in 2.x.

lukelooker commented on packages/components-providers/src/StyleDefender.tsx:
```diff
@@ -24,48 +24,33 @@
-  }
-`
-
-export const GlobalStyle = createGlobalStyle`
-  ${reset}
```
This is the most important change in the entire PR arguably.

lukelooker commented on packages/components/src/Button/ButtonItem.test.tsx:
```diff
@@ -24,31 +24,43 @@
-import React, { FC, useContext } from 'react'
-import { ThemeContext, ThemeProvider } from 'styled-components'
-import { GlobalStyle } from './GlobalStyle'
+import '@testing-library/jest-dom/extend-expect'
```
This is a new file but Git is being goofy and think it's a rename of a completely unrelated (and deleted) file.

lukelooker commented on packages/components-providers/src/GlobalStyle.tsx:
```diff
@@ -24,5 +24,10 @@
+ * @deprecated - This component is no longer needed and is a "no-op"
+ * @todo - Remove in 2.x
+ */
+export const GlobalStyle = () => null
```
No longer used internally but needs to be preserved to prevent a breaking change downstream.

lukelooker commented on packages/components/src/Button/ButtonGroup.test.tsx:
```diff
@@ -115,4 +115,16 @@ describe('ButtonGroup', () => {
     expect(onChangeMock).not.toHaveBeenCalled()
   })
+
+  test('JSX children', () => {
+    renderWithTheme(
+      <ButtonGroup>
+        <ButtonItem>
+          <strong>BOLD</strong> thing
+        </ButtonItem>
+      </ButtonGroup>
+    )
+
+    expect(screen.getByRole('button')).toBeInTheDocument()
+  })
```
Chasing 100% coverage on `ButtonItem` & `ButtonGroup`

dbchristopher commented on packages/components-providers/src/ComponentsProvider.tsx:
```diff
@@ -110,15 +125,17 @@ export const ComponentsProvider: FC<ComponentsProviderProps> = ({
 
   useI18n({ locale, resources })
 
+  const ConditionalStyleDefender = snapshotMode ? Fragment : StyleDefender
```
👍

dbchristopher reviewed (dismissed):
This seems good! I like the lighter touch.

mdodgelooker commented on packages/components/src/Button/ButtonBase.tsx:
```diff
@@ -130,7 +130,9 @@ const ButtonJSX = forwardRef(
 
-export const GenericButtonBase = styled(ButtonJSX)<ButtonProps>``
+export const GenericButtonBase = styled(ButtonJSX)<ButtonProps>`
+  font-family: inherit;
```
Doesn't this get applied via `reset` on ln 43 above? Are you doubling up in case the developer has a `reset` override that's missing it?

lukelooker replied:
I think in earlier revisions of `reset.ts` I didn't have `font-family: inherit` in place.

mdodgelooker commented on packages/components/src/Portal/Portal.tsx:
```diff
@@ -92,7 +93,14 @@ export const Portal = forwardRef(
+ * "CSS reset"
+ */
 const InvisiBox = styled.div<PortalPlacementProps>`
+  ${styleDefenderCSS}
```
The white background blocks out what's underneath.
```suggestion
  ${styleDefenderCSS}
  background: transparent;
```
![de9e81f8-020a-4965-8c22-a07b8180e10b](https://user-images.githubusercontent.com/53451193/115077059-4beb2100-9eb2-11eb-8c5a-aece3f4c0eae.gif)

mdodgelooker replied:
But even with that background style change it seems like somethings off with the `Popover` click outside behavior on `/?path=/story/popover--popover-focus-trap` – with "Cancel Click Outside" off, the "Click me with the popover open" button should trigger an alert. I can dig into it a bit more next week.

mdodgelooker replied:
Please ignore that last comment. It's working normally with or without the background. I must've been trippin.

lukelooker replied:
Move background into `StyleDefender` component instead of share CSS block to address background overlay issue. Realized that I had tested in Dialog/Drawer but not Popover 🤦

mdodgelooker reviewed (changes requested):
Really cool! I have a few questions just out of curiosity, and then I found some `Portal` issues.

lukelooker commented on packages/components-providers/src/StyleDefender.tsx:
```diff
@@ -24,25 +24,29 @@
+export const styleDefenderCSS = css`
+  background: ${({ theme: { colors } }) => colors.background};
+  box-sizing: border-box;
+  color: initial; /* Might better be set to theme.colors.text eventually */
```
Good point. I think I was seeing override CSS altering color and was trying to mitigate that but that's the scenario currently so it seems ideal to leave this as-is for now to limit scope.

lukelooker commented:
@mdodgelooker Addressed your feedback and added in support for `theme.reset` override in case it's needed during integration. I plan to remove that after initial integration if it proves, as expected, to be unneeded.

mdodgelooker reviewed (approved):
LGTM! 🚢
